apis/core/v1alpha2/featuregate_webhook_test.go

```diff
@@ -0,0 +1,95 @@
+				{ObjectMeta: metav1.ObjectMeta{Name: "hundred"}, Spec: FeatureSpec{Activated: false, Discoverable: false, Immutable: true}},
+				{ObjectMeta: metav1.ObjectMeta{Name: "thousand"}, Spec: FeatureSpec{Activated: false, Discoverable: false, Immutable: false}},
+			},
+			featureGateSpec: FeatureGateSpec{
+				Features: []FeatureReference{
+					// Immutable, previously activated, now deactivated (disallowed).
+					{Name: "one", Activate: false},
+					// Immutable, previously deactivated, now activated (disallowed).
+					{Name: "two", Activate: true},
+					// Non-immutable, previously activated, no state change (allowed).
+					{Name: "three", Activate: true},
+					// Non-immutable, previously deactivated, no state change (allowed).
+					{Name: "four", Activate: false},
+					// Unavailable feature, no state change (allowed).
+					{Name: "six", Activate: true},
+					// Unavailable feature, state changed (allowed).
+					{Name: "seven", Activate: false},
+					// Unavailable feature, no state change (allowed).
+					{Name: "eight", Activate: false},
+					// Unavailable feature, no state change (allowed).
+					{Name: "nine", Activate: true},
+					// Unavailable feature, no state change (allowed).
+					{Name: "ten", Activate: true},
+					// Immutable, previously activated, now deactivated (disallowed).
+					{Name: "eleven", Activate: false},
+					// Immutable, previously deactivated, no state change (allowed).
+					// Not explicitly specified in spec.
+					{Name: "twelve", Activate: false},
+					// Immutable, non-discoverable, no state change (allowed).
+					// Not explicitly specified in spec.
+					{Name: "hundred", Activate: false},
+				},
+			},
+			oldObj: &FeatureGate{Status: FeatureGateStatus{
+				ActivatedFeatures:   []string{"one", "three", "eleven"},
+				DeactivatedFeatures: []string{"two", "four", "five", "twelve", "thirteen"},
+			}},
+			want: []string{"one", "two", "eleven"},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			got := computeChangedImmutableFeatures(tc.featureGateSpec, tc.currentFeatures, tc.oldObj)
+			if diff := sliceDiffIgnoreOrder(got, tc.want); diff != "" {
+				t.Errorf("got immutable features %v, want %v, diff: %s", got, tc.want, diff)
+			}
+		})
+	}
+}
+
+// sliceDiffIgnoreOrder returns a human-readable diff of two string slices.
+// Two slices are considered equal when they have the same length and same elements. The order of the elements is
+// ignored while comparing. Nil and empty slices are considered equal.
+//
+// This function is intended to be used in tests for comparing expected and actual values, and printing the diff for
+// users to debug:
+//
+//	if diff := sliceDiffIgnoreOrder(got, want); diff != "" {
+//	    t.Errorf("got: %v, want: %v, diff: %s", got, want, diff)
+//	}
+func sliceDiffIgnoreOrder(a, b []string) string {
```

Comment:
I'm glad you didn't put this function in a common `utils` or `helper` type package. I know we do it all the time, but I like Dave Cheney's recommendation not to, and instead put these functions (and it's ok to copy/duplicate them) in the packages they're used. 

Dave wrote and [explained in Singapore GopherCon 2019](https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_avoid_package_names_like_base_common_or_util),
> ...move the relevant functions into their caller’s package. Even if this involves duplicating some helper code this is better than introducing an import dependency between two packages.

Reply:
Maybe I'm mistaken. I see you did follow the habit of Tanzu Framework/VMware/Kubernetes and put the function in util/cmp/strings/slices.go. In that case, does this `sliceDiffIgnoreOrder` function still need to be here?